### DISCUSSION TITLE
C# LIst 삭제 질문

### DISCUSSION THREAD

**Discussion by fkdl0048** [fkdl0048/Mentoring-Project#5] [Q&A] [answered]
List 관련 코드를 보고 있는데 반복을 돌며 삭제하는 경우 뒤에서부터 삭제하는 모습이 자주 보이는데 이유를 알고 싶습니다..

추측만 하게 되어서 이런 것도 질문해도 될까요..? 너무 자주 쓸데없는 질문하는게 아닌지.. 죄송합니다..(불편하시다면 말씀해주세요!)

```cs
public void RemoveRedundancies()
    {
        // 새 딕셔너리 생성
        Dictionary<EVENT_TYPE, List<IListener>> TmpListeners = new Dictionary<EVENT_TYPE, List<IListener>>();

        // 모든 딕셔너리 항목을 순회한다.
        foreach (KeyValuePair<EVENT_TYPE, List<IListener>> Item in Listeners)
        {
            // 리스트의 모든 리스너 오브젝트를 순회하며 null 오브젝트를 제거한다.
            for (int i = Item.Value.Count - 1; i >= 0; i--)
            {
                if (Item.Value[i].Equals(null))
                    Item.Value.RemoveAt(i);
            }

            // 알림을 받기 위한 항목들만 리스트에 남으면 이 항목들은 임시 딕셔너리에 담는다.
            if (Item.Value.Count > 0)
                TmpListeners.Add(Item.Key, Item.Value);
        }

        Listeners = TmpListeners;
    }
``` 

추측 1번: 동적 배열이라 순회 도중에 크기가 바뀌면 에러가 나서..?
추측 2번..?

**Comment by jongfeel** [comment, marked as answer]
추측 1번이 맞습니다.

List의 경우 remove를 하면 동적으로 크기가 변하므로 문제가 생기게 됩니다.

**Reply by jongfeel** [reply]
추가로 list collection의 item을 삭제할 때 왜 순차적으로 삭제하면 안되는지 
간단하게 코드를 짜보고 테스트 해본 뒤에 정리해 보는 것도 좋을 것 같습니다.

보통 프로그래밍 처음 배우시는 분들의 블로그 글감으로는 좋은 주제이기도 하고요.

**Reply by fkdl0048** [reply]
재밌는 주제를 던져주셔서 글로 한번 작성해봤습니다..! 
[링크](https://fkdl0048.github.io/csharp/cshape_ListRemove/)  

작성하다가 공식 문서에서 궁금한 점이 생겨서 질문드립니다. 

둘 다 유사한 기능을 가진 클래스를 [ArrayList](https://learn.microsoft.com/ko-kr/dotnet/api/system.collections.arraylist?view=net-6.0) 사용할 [List<T>](https://learn.microsoft.com/ko-kr/dotnet/api/system.collections.generic.list-1?view=net-6.0) 지 여부를 결정할 때 클래스는 [List<T>](https://learn.microsoft.com/ko-kr/dotnet/api/system.collections.generic.list-1?view=net-6.0) 대부분의 경우 더 나은 성능을 발휘하며 형식이 안전합니다. 참조 형식이 클래스 형식 T [List<T>](https://learn.microsoft.com/ko-kr/dotnet/api/system.collections.generic.list-1?view=net-6.0) 에 사용되는 경우 두 클래스의 동작은 동일합니다. 그러나 값 형식이 형식 T에 사용되는 경우 구현 및 boxing 문제를 고려해야 합니다.  
(https://learn.microsoft.com/ko-kr/dotnet/api/system.collections.generic.list-1?view=net-6.0)
(https://stackoverflow.com/questions/560184/in-c-are-the-values-in-a-liststruct-boxed)

- 공식문서의 한부분 인데 따로 인터넷에 검색해보니 list에 struct값을 넣는 경우 박싱을 걱정하지 않아도 된다고 하는데 여기서 말하는 boxing 현상은 뭘 말하는 것인지.. 궁금합니다..

**Reply by fkdl0048** [reply]
추가로 리스트를 삭제하는 도중에 다른 곳에서 리스트에 **삽입**을 하는 경우에도 역방향으로 삭제가 더 안전한 이유가 될까요?

**Reply by jongfeel** [reply]
우선 boxing, unboxing은 공식 문서를 보고 이해하시는 게 좋을 거고요.
https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/types/boxing-and-unboxing

간단하게 설명 드리면 값 형식을 객체 형식으로 참조하게 될 때 일어나는 일이라고 보시면 됩니다.

C# 초기 1.0, 1.1 버전에서 generic이 없던 시절에 많이 고려해야 했던 부분이고
지금은 generic type을 지정해서 프로그래밍 하면 크게 문제될 건 없을 겁니다.

**Reply by jongfeel** [reply]
> 추가로 리스트를 삭제하는 도중에 다른 곳에서 리스트에 삽입을 하는 경우에도 역방향으로 삭제가 더 안전한 이유가 될까요?

이것도 문제가 생길 여지는 있습니다.
이건 reverse loop를 돌아도 안전하지 않겠죠.

**Reply by fkdl0048** [reply]
감사합니다..!
